### PR TITLE
Gauge uses Unicode Blocks

### PR DESCRIPTION
**Pull request by DashEightMate**
This pr allows the Gauge to display more detailed progress by using [Unicode block elements](https://en.wikipedia.org/wiki/Block_Elements). The changes don't have any bearing on outward-facing APIs or any other Gauge function. When there is a conflict between the label and the block character, the label character will display as normal, but with the colors from the ratio rounded down.

### CODE REVIEW

File: src/widgets/gauge.rs [hunk 1]
```diff
@@ -129,6 +134,20 @@ impl<'a> Widget for Gauge<'a> {
     }
 }
 
+fn get_unicode_block<'a>(frac: f64) -> &'a str {
+    match (frac*8.0).round() as u16{ //get how many eighths the fraction is closest to
+        1 => "▏",
```
Comment (fdehau) [inline src/widgets/gauge.rs:139]: Those graphemes are usually defined in [src/symbols.rs](https://github.com/fdehau/tui-rs/blob/master/src/symbols.rs) in order to potentially be used by other widgets.

File: src/widgets/gauge.rs [hunk 2]
```diff
@@ -113,6 +113,11 @@ impl<'a> Widget for Gauge<'a> {
                 buf.get_mut(x, y).set_symbol(" ");
             }
 
+            //set unicode block
+            if self.ratio < 1.0{
+                buf.get_mut(end, y).set_symbol(get_unicode_block(width%1.0));
```
Comment (fdehau) [inline src/widgets/gauge.rs:118]: I would add an option to use the old low resolution behavior as it is always wanted in environments where those unicode symbols are not well supported.

**Review by fdehau — CHANGES_REQUESTED** [review]
It looks like a nice improvement. I've left some initial comments. I would like to see some tests too :).

**Comment by DashEightMate** [comment]
Just made the gauge use the symbols in symbols.rs, and made an option to not use unicode. There's now tests for both with and without unicode, and the bottom gauge on the gauge example has unicode disabled as well.

File: src/widgets/gauge.rs (hunk 1 above)
Comment (DashEightMate) [reply src/widgets/gauge.rs]: Now references the symbols in symbols.rs as of the latest commit

File: src/widgets/gauge.rs (hunk 2 above)
Comment (DashEightMate) [reply src/widgets/gauge.rs]: Latest commit adds an option to revert to the old behavior, and I added a test for both new and old behavior

File: tests/widgets_gauge.rs [hunk 3]
```diff
@@ -28,6 +28,46 @@ fn widgets_gauge_renders() {
             f.render_widget(gauge, chunks[1]);
         })
         .unwrap();
+    let expected = Buffer::with_lines(vec![
+        "                                        ",
+        "                                        ",
+        "  ┌Percentage────────────────────────┐  ",
+        "  │              ▋43%                │  ",
```
Comment (fdehau) [inline tests/widgets_gauge.rs:35]: Nice, could we also check the color behavior ? Especially when we are at 50% and the label may have different colors.

File: src/widgets/gauge.rs [hunk 4]
```diff
@@ -129,6 +154,21 @@ impl<'a> Widget for Gauge<'a> {
...
+    match (frac * 8.0).round() as u16 {
+        //get how many eighths the fraction is closest to
+        1 => block::ONE_EIGHTH,
+        2 => block::ONE_QUARTER,
+        3 => block::THREE_EIGHTHS,
+        4 => block::HALF,
+        5 => block::FIVE_EIGHTHS,
+        6 => block::THREE_QUARTERS,
+        7 => block::SEVEN_EIGHTHS,
+        8 => "",
```
Comment (fdehau) [inline src/widgets/gauge.rs:167]: shouldn't this be a full block ?

File: src/widgets/gauge.rs [hunk 5]
```diff
@@ -81,6 +84,11 @@ impl<'a> Gauge<'a> {
         self.gauge_style = style;
         self
     }
+
+    pub fn use_unicode(mut self, unicode: bool) -> Gauge<'a> {
+        self.use_block = unicode;
```
Comment (fdehau) [inline src/widgets/gauge.rs:89]: Any reason to have the property name be different from the name of the method setting its value ?

File: tests/widgets_gauge.rs (hunk 3 above)
Comment (DashEightMate) [reply tests/widgets_gauge.rs:35]: I'm not sure how to get the color from the terminal/backend for an assert. Is there an example of this somewhere else?

File: src/widgets/gauge.rs (hunk 4 above)
Comment (DashEightMate) [reply src/widgets/gauge.rs]: There isn't a difference, I changed it over so it makes more sense with readability

File: src/widgets/gauge.rs (hunk 5 above)
Comment (DashEightMate) [reply src/widgets/gauge.rs]: Nope, just changed it over

File: tests/widgets_gauge.rs (hunk 3 above)
Comment (fdehau) [reply tests/widgets_gauge.rs:35]: You can simply modify the cell of the expected `Buffer`. See [this](https://github.com/fdehau/tui-rs/blob/5a9b59866bbe93dad00309e4ffc449a21601972d/tests/widgets_list.rs#L33) for an example.

**Comment by fdehau** [comment]
@DashEightMate Sorry for keeping you waiting. Beside the test on colors (I answered your comment about that) and the merge conflict, everything looks good. I would be more than happy to merge this if you are still up for it :).

**Comment by DashEightMate** [comment]
@fdehau No worries! Just added color checks to the gauge as well as merged with main. Let me know if this needs anything else.

**Pull request merged**